sieve_of_eratosthenes/SieveOfEratosthenes.java

```diff
@@ -1,34 +1,128 @@
+ *
+ * <p>When creating a Sieve, you need to specify the max number that you might query.</p>
+ *
+ * <p>Time complexity for creating the Sieve is O(n * n) but for querying is O(1). Space complexity is O(n)</p>
+ *
+ * @see <a href="https://en.wikipedia.org/wiki/Sieve_of_Eratosthenes">Sieve of Eratosthenes</a>
+ */
 public class SieveOfEratosthenes {
 
-    public static boolean[] sieveOfEratosthenes(int n) {
-        int sqrtOfn = (int)Math.sqrt(n) + 1;
-        boolean[] primes = new boolean[n + 1];
-        for (int i = 2; i < n + 1; ++i) {
-            primes[i] = true;
-        }
-        for (int i = 2; i <= sqrtOfn; ++i) {
-            if (primes[i]) {
-                for (int j = 2 * i; j < n + 1; j += i) {
-                    primes[j] = false;
-                }
+    private boolean[] sieve;
+
+    /**
+     * Creates a sieve that can be queried for numbers up to maxNumber.
+     *
+     * @param maxNumber
+     */
+    public SieveOfEratosthenes(final int maxNumber) {
+        initSieve(maxNumber);
+        uncrossAllNumbers();
+        crossObviousNumbers();
+        crossMultiplesOfKnownPrimes();
+    }
+
+    /**
+     * Iterates the sieve and when it finds a prime in it, it'll cross its multiples, since by definition those numbers are not prime.
+     */
+    private void crossMultiplesOfKnownPrimes() {
+        for (int n = 2; nextMultiplePossiblyNotCrossedOf(n) < sieve.length; ++n) {
+            if (isPrime(n)) {
+                crossMultiplesOf(n);
             }
         }
-        return primes;
+    }
+
+    /**
+     * Given prime, crosses all multiples of prime in the sieve
+     *
+     * @param prime
+     */
+    private void crossMultiplesOf(final int prime) {
+        for (int n = nextMultiplePossiblyNotCrossedOf(prime); n < sieve.length; n = nextMultipleOf(prime, n)) {
+            cross(n);
+        }
+    }
+
+    /**
+     * Crosses 0 and 1 which by definition aren't prime.
+     */
+    private void crossObviousNumbers() {
+        cross(0);
+        cross(1);
+    }
+
+    /**
+     * Returns the next multiple of n which is greater than last.
+     *
+     * @param n
+     * @param last
+     * @return
+     */
+    private int nextMultipleOf(final int n, final int last) {
+        return n + last;
+    }
+
+    /**
+     * <p>Returns the next number which is multiple of n, which possibly hasn't been crossed</p>
+     *
+     * <p>n * 2 has been cross, while crossing the multiples of 2, n * 3 while crossing the multiple of 3, and so on.
+     * So the first multiple that possibly hasn't been crossed is n * n.</p>
+     *
+     * @param n
+     * @return
+     */
+    private int nextMultiplePossiblyNotCrossedOf(final int n) {
+        return n * n;
+    }
+
+    /**
+     * Initializes the sieve to be queried to numbers up to maxNumber.
+     *
+     * @param maxNumber
+     */
+    private void initSieve(final int maxNumber) {
+        this.sieve = new boolean[maxNumber + 1];
+    }
```

Comment:
This can go to the constructor.

Reply:
@singhpratyush this method is used to hide the fact that the sieve is an array. If you see, there's a lot of methods that you can check and understand what's the algorithm doing without you realizing there's an array backing the sieve. That means that you don't mix domain details with implementation details which help to make the code easier to understand.